Sample中的DynamicRuntime.recoveryRuntime(application)方法Crash

1. 先修复InitApplication中isProcess的return 错误： return currentProcName.endsWith(processName);
2. 运行maven下的sample，必crash :
2019-11-20 15:00:09.224 14459-14459/com.tencent.midas:plugin D/AndroidRuntime: Shutting down VM
    
    
    --------- beginning of crash
2019-11-20 15:00:09.225 14459-14459/com.tencent.midas:plugin E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.tencent.midas:plugin, PID: 14459
    java.lang.ExceptionInInitializerError
        at com.tencent.shadow.dynamic.host.DynamicRuntime.recoveryRuntime(DynamicRuntime.java:175)
        at com.tencent.midassdk.api.MidasAPI.init(MidasAPI.java:44)
        at com.tencent.midas.MyApplication.onCreate(MyApplication.java:13)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1189)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6460)
        at android.app.ActivityThread.access$1300(ActivityThread.java:219)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1859)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
     Caused by: java.lang.RuntimeException: 没有找到 ILoggerFactory 实现，请先调用setILoggerFactory
        at com.tencent.shadow.core.common.LoggerFactory.getILoggerFactory(LoggerFactory.java:39)
        at com.tencent.shadow.core.common.LoggerFactory.getLogger(LoggerFactory.java:33)
        at com.tencent.shadow.dynamic.host.DynamicRuntime.<clinit>(DynamicRuntime.java:43)
        at com.tencent.shadow.dynamic.host.DynamicRuntime.recoveryRuntime(DynamicRuntime.java:175) 
        at com.tencent.midassdk.api.MidasAPI.init(MidasAPI.java:44) 
        at com.tencent.midas.MyApplication.onCreate(MyApplication.java:13) 
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1189) 
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6460) 
        at android.app.ActivityThread.access$1300(ActivityThread.java:219) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1859) 
        at android.os.Handler.dispatchMessage(Handler.java:107) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.app.ActivityThread.main(ActivityThread.java:7356) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930) 

--- a/projects/sample/maven/host-project/introduce-shadow-lib/src/main/java/com/tencent/shadow/sample/introduce_shadow_lib/InitApplication.java
+++ b/projects/sample/maven/host-project/introduce-shadow-lib/src/main/java/com/tencent/shadow/sample/introduce_shadow_lib/InitApplication.java
@@ -52,7 +52,7 @@ public class InitApplication {
     }
 
     private static boolean isProcess(Context context, String processName) {
-        String currentProcName = null;
+        String currentProcName = "";
         ActivityManager manager =
                 (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
         for (ActivityManager.RunningAppProcessInfo processInfo : manager.getRunningAppProcesses()) {
@@ -61,6 +61,7 @@ public class InitApplication {
                 break;
             }
         }
-        return processName.equals(currentProcName);
+
+        return currentProcName.endsWith(processName);
     }
 }

--- a/projects/sample/maven/host-project/introduce-shadow-lib/src/main/java/com/tencent/shadow/sample/introduce_shadow_lib/InitApplication.java
+++ b/projects/sample/maven/host-project/introduce-shadow-lib/src/main/java/com/tencent/shadow/sample/introduce_shadow_lib/InitApplication.java
@@ -25,15 +25,15 @@ public class InitApplication {
     }
 
     public static void onApplicationCreate(Application application) {
+        //Log接口Manager也需要使用，所以主进程也初始化。
+        LoggerFactory.setILoggerFactory(new AndroidLoggerFactory());
+
         if (isProcess(application, ":plugin")) {
             //在全动态架构中，Activity组件没有打包在宿主而是位于被动态加载的runtime，
             //为了防止插件crash后，系统自动恢复crash前的Activity组件，此时由于没有加载runtime而发生classNotFound异常，导致二次crash
             //因此这里恢复加载上一次的runtime
             DynamicRuntime.recoveryRuntime(application);
         }
-
-        //Log接口Manager也需要使用，所以主进程也初始化。
-        LoggerFactory.setILoggerFactory(new AndroidLoggerFactory());
 
         FixedPathPmUpdater fixedPathPmUpdater
                 = new FixedPathPmUpdater(new File("/data/local/tmp/sample-manager-debug.apk"));

--- a/projects/sample/source/sample-host/src/main/java/com/tencent/shadow/sample/host/HostApplication.java
+++ b/projects/sample/source/sample-host/src/main/java/com/tencent/shadow/sample/host/HostApplication.java
@@ -18,7 +18,9 @@
 
 package com.tencent.shadow.sample.host;
 
+import android.app.ActivityManager;
 import android.app.Application;
+import android.content.Context;
 import android.os.Build;
 import android.os.StrictMode;
 
@@ -29,6 +31,8 @@ import com.tencent.shadow.sample.host.lib.HostUiLayerProvider;
 import com.tencent.shadow.sample.host.manager.Shadow;
 
 import java.io.File;
+
+import static android.os.Process.myPid;
 
 public class HostApplication extends Application {
     private static HostApplication sApp;
@@ -44,10 +48,12 @@ public class HostApplication extends Application {
 
         LoggerFactory.setILoggerFactory(new AndroidLogLoggerFactory());
 
-        //在全动态架构中，Activity组件没有打包在宿主而是位于被动态加载的runtime，
-        //为了防止插件crash后，系统自动恢复crash前的Activity组件，此时由于没有加载runtime而发生classNotFound异常，导致二次crash
-        //因此这里恢复加载上一次的runtime
-        DynamicRuntime.recoveryRuntime(this);
+        if (isProcess(this, ":plugin")) {
+            //在全动态架构中，Activity组件没有打包在宿主而是位于被动态加载的runtime，
+            //为了防止插件crash后，系统自动恢复crash前的Activity组件，此时由于没有加载runtime而发生classNotFound异常，导致二次crash
+            //因此这里恢复加载上一次的runtime
+            DynamicRuntime.recoveryRuntime(this);
+        }
 
         PluginHelper.getInstance().init(this);
 
@@ -76,5 +82,19 @@ public class HostApplication extends Application {
 
     public PluginManager getPluginManager() {
         return mPluginManager;
+    }
+
+    private static boolean isProcess(Context context, String processName) {
+        String currentProcName = "";
+        ActivityManager manager =
+                (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
+        for (ActivityManager.RunningAppProcessInfo processInfo : manager.getRunningAppProcesses()) {
+            if (processInfo.pid == myPid()) {
+                currentProcName = processInfo.processName;
+                break;
+            }
+        }
+
+        return currentProcName.endsWith(processName);
     }
 }

--- a/projects/sample/sunflower/host-project/introduce-shadow-lib/src/main/java/com/tencent/shadow/sample/introduce_shadow_lib/InitApplication.java
+++ b/projects/sample/sunflower/host-project/introduce-shadow-lib/src/main/java/com/tencent/shadow/sample/introduce_shadow_lib/InitApplication.java
@@ -52,7 +52,7 @@ public class InitApplication {
     }
 
     private static boolean isProcess(Context context, String processName) {
-        String currentProcName = null;
+        String currentProcName = "";
         ActivityManager manager =
                 (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
         for (ActivityManager.RunningAppProcessInfo processInfo : manager.getRunningAppProcesses()) {
@@ -61,6 +61,7 @@ public class InitApplication {
                 break;
             }
         }
-        return processName.equals(currentProcName);
+
+        return currentProcName.endsWith(processName);
     }
 }

--- a/projects/sample/sunflower/host-project/introduce-shadow-lib/src/main/java/com/tencent/shadow/sample/introduce_shadow_lib/InitApplication.java
+++ b/projects/sample/sunflower/host-project/introduce-shadow-lib/src/main/java/com/tencent/shadow/sample/introduce_shadow_lib/InitApplication.java
@@ -25,15 +25,15 @@ public class InitApplication {
     }
 
     public static void onApplicationCreate(Application application) {
+        //Log接口Manager也需要使用，所以主进程也初始化。
+        LoggerFactory.setILoggerFactory(new AndroidLoggerFactory());
+
         if (isProcess(application, ":plugin")) {
             //在全动态架构中，Activity组件没有打包在宿主而是位于被动态加载的runtime，
             //为了防止插件crash后，系统自动恢复crash前的Activity组件，此时由于没有加载runtime而发生classNotFound异常，导致二次crash
             //因此这里恢复加载上一次的runtime
             DynamicRuntime.recoveryRuntime(application);
         }
-
-        //Log接口Manager也需要使用，所以主进程也初始化。
-        LoggerFactory.setILoggerFactory(new AndroidLoggerFactory());
 
         FixedPathPmUpdater fixedPathPmUpdater
                 = new FixedPathPmUpdater(new File("/data/local/tmp/sample-manager-debug.apk"));

